packages/ember-routing/lib/system/router_state.js

```diff
@@ -11,11 +11,8 @@ export default EmberObject.extend({
     let state = this.routerJsState;
     if (!this.routerJs.isActiveIntent(routeName, models, null, state)) { return false; }
 
-    let emptyQueryParams = isEmpty(Object.keys(queryParams));
-
-    if (queryParamsMustMatch && !emptyQueryParams) {
-      let visibleQueryParams = {};
-      assign(visibleQueryParams, queryParams);
+    if (queryParamsMustMatch && Object.keys(queryParams).length > 0) {
```

Comment:
cannot think of clean way doing it, at least doesn't check length until queryParamsMustMatch is true, PR intention was to just reuse `shallowEqual` this is just what I picked up on the way